### PR TITLE
example of post-processor to have integer CPU

### PR DESCRIPTION
**Pull request by dbenque**
#### Which component this PR applies to?

vertical-pod-autoscaler/recommender

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

It introduces a recommendation post processor that allows users to have a CPU integer as recommendation.
Thanks to that it is now possible to VPA for application using [static CPU allocation](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy)

#### Which issue(s) this PR fixes:

Fixes #5236

#### Special notes for your reviewer:

This PR is an example about how to use #5239

#### Does this PR introduce a user-facing change?

It does not break any existing API.
With this implementation users activate the feature by adding an annotation their VPA objects:
`vpa-post-processor.kubernetes.io/{containerName}_integerCPU=true`

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

This post processor is optional. To run it in the recommender the command-line flag should be set:
`--recommendation-post-processors="integerCPU,capping"`

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/main.go [hunk 1]
```diff
@@ -57,6 +57,9 @@ var (
 	ctrPodNameLabel     = flag.String("container-pod-name-label", "pod_name", `Label name to look for container pod names`)
 	ctrNameLabel        = flag.String("container-name-label", "name", `Label name to look for container names`)
 	vpaObjectNamespace  = flag.String("vpa-object-namespace", apiv1.NamespaceAll, "Namespace to search for VPA objects and pod stats. Empty means all namespaces will be used.")
+
+	// recommendation post processor flags
+	postProcessorCPUasInteger = flag.Bool("recommendation-post-processor-cpu-as-integer-enabled", false, `Enable recommendation post process to have CPU as integer if requested by user in VPA object`)
```
Comment (iksaif) [inline vertical-pod-autoscaler/pkg/recommender/main.go:62]: I think this should be a list of pre processors instead

File: vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// integerCPUPostProcessor ensure that the policy is applied to recommendation
+// it applies policy for fields: cpuAsInteger
+type integerCPUPostProcessor struct{}
+
+const (
+	// The user interface for that post processor is an annotation with the following format:
+	// vpa-post-processor.kubernetes.io/{containerName}_integerCPU=true
+
+	vpaPostProcessorPrefix           = "vpa-post-processor.kubernetes.io/"
+	vpaPostProcessorIntegerCPUSuffix = "_integerCPU"
```
Comment (iksaif) [inline vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go:36]: shouldn't this be `-integer-cpu` ?

File: vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+limitations under the License.
+*/
+
+package routines
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+	"math"
```
Comment (iksaif) [inline vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go:23]: import ordering seems wrong

File: vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+func (c integerCPUPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, policy *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+
+	amendedRecommendation := recommendation.DeepCopy()
+
+	process := func(recommendation apiv1.ResourceList) {
+		for resourceName, recommended := range recommendation {
+			if resourceName != apiv1.ResourceCPU {
+				continue
+			}
+			v := float64(recommended.MilliValue()) / 1000
```
Comment (iksaif) [inline vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go:52]: `1000.` ?

File: vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			process(r.UpperBound)
+			process(r.UncappedTarget)
+		}
+	}
+	return amendedRecommendation
+}
+
+// extractContainerName return the container name for the feature based on annotation key
+// if the return value is empty that means that the key does not match
+func extractContainerName(key, prefix, suffix string) string {
```
Comment (iksaif) [inline vertical-pod-autoscaler/pkg/recommender/routines/cpu_integer_post_processor.go:80]: This is likely to be re-used and should be in a shared module no ?

**Comment by dbenque** [comment]
Merged upstream

**Pull request closed without merging**